cpp/ql/lib/semmle/code/cpp/ir/dataflow/internal/DataFlowUtil.qll

```diff
@@ -591,10 +596,14 @@ class InitialGlobalValue extends Node, TInitialGlobalValue {
 
   override Declaration getFunction() { result = globalDef.getIRFunction().getFunction() }
 
+  final override predicate isGLValue() { globalDef.getIndirectionIndex() = 0 }
+
   override DataFlowType getType() {
-    exists(int indirectionIndex |
-      indirectionIndex = globalDef.getIndirectionIndex() and
-      result = getTypeImpl(globalDef.getUnspecifiedType(), indirectionIndex)
+    exists(DataFlowType type |
+      type = globalDef.getUnspecifiedType() and
+      if this.isGLValue()
+      then result = globalDef.getUnspecifiedType()
+      else result = getTypeImpl(type.getUnspecifiedType(), globalDef.getIndirectionIndex() - 1)
```

Comment:
Since we have `type = globalDef.getUnspecifiedType()`, I think the following should do.
```suggestion
      if this.isGLValue()
      then result = type
      else result = getTypeImpl(type, globalDef.getIndirectionIndex() - 1)
```

Reply:
What I don't understand is why here we do not recurse in the case of a glvalue, while in the two case below we do.

Reply:
> Since we have type = globalDef.getUnspecifiedType(), I think the following should do.

Agreed. Thanks for that!

Reply:
Yep. I'm trying to formulate a good response to this right now 😂.

Reply:
Take your time, I though it might have slipped through, but apparently it didn't.

Reply:
> What I don't understand is why here we do not recurse in the case of a glvalue, while in the two case below we do.

Yeah, that's a bit unfortunate. I couldn't quite make it fit the pattern as the other cases. The reason is that, the other cases (i.e., `IndirectOperand` and `IndirectInstruction`) are defined with indirection indices in the range 1 to _some max number_.

However, the `GlobalDef` that defined an `InitialGlobalValue` has an index that's in the range 0 to _some max number_. And that discrepancy between the ranges meant I couldn't figure out how to write it in the same style as the other two cases 😕.

It may be that the real fix is to modify the indices of `InitialGlobalValue` so that its indirection index range starts 1 instead of 0 🤔. However, I'd like to delay such a larger change to another PR if you're okay with it. I can also add a version of the above explanation to a comment in this predicate if you think that's a good idea.

Reply:
So, `isGLValue` can only be true is the index has the minimum value?

Reply:
`isGLValue` is only implemented for `InstructionNode`s, `OperandNode`s and `InitialGlobalValue`.`InstructionNode` and `OperandNode` (which intuitively can be seen as `IndirectInstructionNode`s and `IndirectOperandNode`s with indirection index 0) implement the predicate by checking if the underlying operand or instruction is a glvalue. Similarly, `InitialGlobalValue` implements the glvalue check by checking if the indirection index is 0.

Does that make sense? I'd be happy to do a Zoom chat about it if that's any easier.